autogluon/utils/tabular/features/types.py

```diff
@@ -0,0 +1,126 @@
+        return 'bool'
+    elif dtype.name in ['str', 'string', 'object']:
+        return 'object'
+    else:
+        return dtype.name
+
+
+# Real dtypes
+def get_type_map_real(df: DataFrame) -> dict:
+    features_types = df.dtypes.to_dict()
+    return {k: v.name for k, v in features_types.items()}
+
+
+# Raw dtypes (Real dtypes family)
+def get_type_map_raw(df: DataFrame) -> dict:
+    features_types = df.dtypes.to_dict()
+    return {k: get_type_family_raw(v) for k, v in features_types.items()}
+
+
+def get_type_map_special(X: DataFrame) -> dict:
+    type_map_special = {}
+    for column in X:
+        type_special = get_type_special(X[column])
+        if type_special is not None:
+            type_map_special[column] = type_special
+    return type_map_special
+
+
+def get_type_special(X: Series) -> str:
+    if check_if_datetime_as_object_feature(X):
+        type_special = 'datetime_as_object'
+    elif check_if_nlp_feature(X):
+        type_special = 'text'
+    else:
+        type_special = None
+    return type_special
+
+
+def get_type_group_map(type_map: dict) -> defaultdict:
+    type_group_map = defaultdict(list)
+    for key, val in type_map.items():
+        type_group_map[val].append(key)
+    return type_group_map
+
+
+def get_type_group_map_real(df: DataFrame) -> defaultdict:
+    type_map_real = get_type_map_real(df)
+    return get_type_group_map(type_map_real)
+
+
+def get_type_group_map_raw(df: DataFrame) -> defaultdict:
+    type_map_raw = get_type_map_raw(df)
+    return get_type_group_map(type_map_raw)
+
+
+# TODO: Expand to enable multiple special types per feature
+def get_type_group_map_special(df: DataFrame) -> defaultdict:
+    type_map_special = get_type_map_special(df)
+    return get_type_group_map(type_map_special)
+
+
+# TODO: Expand to int64 -> date features (milli from epoch etc)
+def check_if_datetime_as_object_feature(X: Series) -> bool:
+    type_family = get_type_family_raw(X.dtype)
+    # TODO: Check if low numeric numbers, could be categorical encoding!
+    # TODO: If low numeric, potentially it is just numeric instead of date
+    if X.isnull().all():
+        return False
+    # if type_family == 'datetime':
+    #     return True
+    if type_family != 'object':  # TODO: seconds from epoch support
+        return False
+    try:
+        X.apply(pd.to_numeric)
+    except:
+        try:
+            X.apply(pd.to_datetime)
+            return True
+        except:
+            return False
+    else:
+        return False
+
+
+def check_if_nlp_feature(X: Series) -> bool:
+    type_family = get_type_family_raw(X.dtype)
+    if type_family != 'object':
+        return False
+    X_unique = X.unique()
+    num_unique = len(X_unique)
+    num_rows = len(X)
+    unique_ratio = num_unique / num_rows
+    if unique_ratio <= 0.01:
+        return False
+    avg_words = np.mean([len(re.sub(' +', ' ', value).split(' ')) if isinstance(value, str) else 0 for value in X_unique])
```

Comment:
This might be very expensive on large datasets (especially with long strings) because of regexps. I think this vectorized version will both: work faster and consume less memory:
```python
avg_words = X_unique.str.split().str.len().fillna(0).mean()
```

Here's how it handles extra spaces:
```python
pd.DataFrame({'col': ['a b c','a  b  c','a    b    c', np.NaN]})['col'].str.split().str.len().fillna(0)
```
Out:
```
0    3.0
1    3.0
2    3.0
3    0.0
Name: col, dtype: float64
```

Reply:
Thanks for this! I did a test and this is about 2.5x faster, and it handles edge cases better too regarding starting/ending spaces. I have also removed the .fillna(0) part as it is slightly better to just ignore nans (should basically never make a difference regardless).

On my laptop, this made 50,000 rows of paragraph length text run this function in 1 seconds from 2.5 seconds.
Another interesting observation: 50,000 rows of datetime_as_object takes 20-30 seconds in the check_if_datetime_as_object function, so I've added a TODO to optimize that.